EEx: Add trim option

oneeman opened the pull request:
If the `trim` option is used, trims around EEx tags:
- On left, up to the line break if all whitespace
- On right, up to and including the line break if all whitespace

Closes #3154

Initial attempt.  I think the behavior is ok, at least if I understood the idea correctly.  The only concern is that it also trims after `%>` even if it's not part of a tag (e.g. `%>  \nfoo` becomes `%>foo`).  This is a consequence of how the quotations are handled, and I don't know of a way to avoid it without redoing that using delimiter balancing.  If it's a problem, I could give it more thought.  Otherwise, some of the choices made:

- The trim option is passed inside `opts` rather than as a boolean flag (I figured this way it'd be easier to add more options later on)
- For whitespace it's tabs or spaces
- For newlines it's checking both LF and CRLF

I wasn't 100% sure about these, but I figured I might as well go ahead and I can change whatever is needed afterwards.

oneeman commented:
Rerunning the build after (probably unrelated) failure: https://travis-ci.org/elixir-lang/elixir/jobs/55436354

josevalim commented:
> The only concern is that it also trims after %> even if it's not part of
a tag (e.g. %> \nfoo becomes %>foo).

What do you mean? Do we have an example in the suite? I thought %> is
always part of the tag (exactly because we don't balance?).

Also, I think we don't want to trim on quotations (because the trim will
happen when the quotation is evaluated). Not sure how doable this is. If
not in the tokenizer, we would have to move it to the parser/compiler.

Thanks!


-- 


*José Valim*
www.plataformatec.com.br
Skype: jv.ptec
Founder and Lead Developer

oneeman commented:
On master:
```
iex(1)> EEx.eval_string "\n  foo %>  \nbar"
"\n  foo %>  \nbar"
```

On eex-trim (before I added the second commit):
```
iex(1)> EEx.eval_string "\n  foo %>  \nbar", [], trim: true
"\n  foo %>bar"
```

This was happening because quotations are now done without calling `expr`, so I had to add a function clause to `tokenize` to trim after `%>`, and it couldn't know whether the `%>` was ending a quotation or stand-alone like in this example.  So, not trimming around quotations actually eliminates this problem, and so I went ahead and added that commit.

On eex-trim now:
```
iex(1)> EEx.eval_string "\n  foo %>  \nbar", [], trim: true
"\n  foo %>  \nbar"
```

I haven't added this to 'no false positives' test because I wasn't sure if this was intentional (i.e. whether we officially support stand-alone `%>`s or whether it's supported merely because we don't throw an error on it).  I can add it as a case if that's appropriate.

oneeman commented:
Another thing, which I just noticed: it does trim around any tags inside a quotation:

```
iex(1)> string = "<%% foo\n      <%= 10 %>   \n%>"
"<%% foo\n      <%= 10 %>   \n%>"
iex(2)> EEx.eval_string(string, [], trim: true)
"<% foo\n10%>"
```

This is because the quotations aren't handled as a whole, just their opening tag outputs `<%` instead and the tokenizer goes back to normal.  This is probably not a big problem if the quotations are evaluated later on, although it could mess up the line numbers.  I can think of two ways if we need to:

(a) Using `expr` for quotations, but that would require balancing delimiters
(b) Having `tokenizer` keep track of whether you're in a quotation or not using an additional argument or something like that, and if it encounters a tag while inside a quotation have it tokenize that tag with `trim: false`.  It would get trickier with quotations inside quotations (we'd essentially be balancing delimiters, but only `<%%` and `%>` outside of non-quotation tags).

Neither of these sound too great, though...

josevalim commented:
> I haven't added this to 'no false positives' test because I wasn't sure if this was intentional (i.e. whether we officially support stand-alone %>s or whether it's supported merely because we don't throw an error on it). I can add it as a case if that's appropriate.

Another good catch, thanks. Ruby's ERb does not raise in this case... so I am unsure if we should. If we do, we need to support some sort of escaping for closing tags too which may be reason to keep today's current behaviour.

josevalim commented:
> Another thing, which I just noticed: it does trim around any tags inside a quotation:

That is fine in my opinion. I plan to use those new features to clean up Phoenix generators, I will let you know how it goes. :)

oneeman commented:
Awesome!  I squashed the two commits and updated the commit message with some more details.  Let me know if anything else needs to be done, otherwise it's ready to merge.

josevalim commented:
I have tried this with Phoenix and this looks fantastic. Thank you! There is just one change I would like to discuss and I would appreciate your opinion:

https://github.com/phoenixframework/phoenix/blob/ec64c773382a610e89f4dc81ce5d3c1c71131140/priv/templates/new/config/dev.exs#L14

The line above is eating the following new line, generating wrong output. To fix it, I would need to add two new lines after `<% end %>`. I think the proper fix is that `%>` should only trim the new line on the right side if the left side was also trimmed up to the new line. I.e. the trimming only happens if we have a single expression, and no text, in that line.

Do you think it would be possible to have such change? Does it make sense?

oneeman commented:
Sounds good.  In fact, the division of trimming on the left and on the right is no longer needed now that we're not trimming around quotations, I expect doing it to be straightforward, and the code added to `tokenize` will be just one call instead of one for each side as it is now.

There is one possible complication, though.  Trimming on the left is done only if encountering a newline.  As a result, the first line of a template would never be trimmed.  For example:
```
<%= if true do %>
YES
<% else %>
NO 
<% end %>
```

would lead evaluate to "\nYES\n", whereas presumably we would want "YES\n".  With the PR in its current state, the first line would be trimmed on the right but not on the left (which is less of an issue, since the first line normally wouldn't start with whitespace).  Anyway, supporting trimming on the first line won't be too complicated.  I just wanted to mention it to confirm that that is the desired behavior.

josevalim commented:
Ah, definitely. We should trim on the right side if the left side is a new line or the beginning of a file. Good catch!

oneeman commented:
Added a commit implementing the changes and redoing the test cases somewhat.  As expected, supporting trimming on the first (and also on the last) line adds a bit of complexity, but nothing too bad... and it does make for some cleaner tests as well.  Will squash if this works well in practice.

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
